moveit_core/online_signal_smoothing/src/acceleration_filter.cpp

```diff
@@ -0,0 +1,145 @@
+{
+  node_ = node;
+  num_joints_ = num_joints;
+
+  cur_velocity_ = Eigen::VectorXd::Zero(num_joints);
+  prev_velocity_ = Eigen::VectorXd::Zero(num_joints);
+  cur_acceleration = Eigen::VectorXd::Zero(num_joints);
+
+  auto param_listener_ = online_signal_smoothing::ParamListener(node_);
+  auto params = param_listener_.get_params();
+  joint_acceleration_limit_ = params.joint_acceleration_limit;
+  update_timeout_ = params.update_timeout;
+
+  return true;
+}
+
+bool AccelerationLimitedPlugin::doSmoothing(Eigen::VectorXd& positions, Eigen::VectorXd& /* unused */,
+                                            Eigen::VectorXd& /* unused */)
+{
+  const long num_positions = positions.size();
+  if (num_positions != num_joints_)
+  {
+    RCLCPP_ERROR_THROTTLE(
+        node_->get_logger(), *node_->get_clock(), 1000,
+        "The length of the joint positions parameter is not equal to the number of joints, expected %zu got %zu.",
+        num_joints_, num_positions);
+    return false;
+  }
+  else if (last_positions_[0].first.size() != num_joints_)
+  {
+    RCLCPP_ERROR_THROTTLE(node_->get_logger(), *node_->get_clock(), 1000,
+                          "Reset must be called before doSmoothing for the AccelerationLimitedPlugin plugin.");
+    return false;
+  }
+
+  if (!last_positions_[0].second.has_value())
+  {
+    positions = last_positions_[0].first;
+    last_positions_[0].second = node_->now();
+  }
+  else if (!last_positions_[1].second.has_value())
+  {
+    positions = last_positions_[1].first;
+    last_positions_[1].second = node_->now();
+  }
+  else
+  {
+    rclcpp::Time cur_time = node_->now();
+    double dt_1 = (last_positions_[1].second.value() - last_positions_[0].second.value()).seconds();
+    double dt_2 = (cur_time - last_positions_[1].second.value()).seconds();
+    if (dt_1 > update_timeout_ || dt_2 > update_timeout_)
+    {
+      RCLCPP_ERROR_THROTTLE(node_->get_logger(), *node_->get_clock(), 1000,
+                            "The maximum allowed time between filter updates exceeded! The `doSmoothing` method must "
+                            "be called at a higher rate.");
+      return false;
+    }
+
+    prev_velocity_ = (last_positions_[1].first - last_positions_[0].first) / dt_1;
+    cur_velocity_ = (positions - last_positions_[1].first) / dt_2;
+    cur_acceleration = (cur_velocity_ - prev_velocity_) / dt_2;
+    double scale = 1.0;
+    for (long i = 0; i < num_joints_; i++)
+    {
+      scale = std::min(scale, std::clamp(cur_acceleration[i], -joint_acceleration_limit_, joint_acceleration_limit_) /
+                                  cur_acceleration[i]);
+    }
+    cur_acceleration *= scale;
+    cur_velocity_ = prev_velocity_ + cur_acceleration * dt_2;
+    positions = last_positions_[1].first + cur_velocity_ * dt_2;
+
+    last_positions_[0] = last_positions_[1];
+    last_positions_[1].first = positions;
+    last_positions_[1].second = cur_time;
+  }
```

Comment:
This whole block of code, which has the actual math, could use some more comments to describe what's going on.